### PR TITLE
OSD-11284 : Add org related ocm-one line commands 

### PR DESCRIPTION
**Pull request by samanthajayasinghe**
### Introduction 
This pull request extends `osdctl` commands to get the organization information. More information is described on [OSD-11284](https://issues.redhat.com/browse/OSD-11284 ) issue. 

### How to use it 
This pull request introduces the following `org` commands. 

#### What's my organization?

The following command shows the current organization. 
```
$osdctl org current 

sample output 
=======================================
ID:              AAAABBBBCCCC
Name:            My org name
External ID:     111111
EBS ID:          2222222
Created:         2019-02-15T20:26:12.542449Z
Updated:         2022-12-21T01:16:25.796805Z
======================================

$osdctl org current  -o json

Json output
=======================================
{
  "id": "AAAABBBBCCCC",
  "external_id": "111111",
  "name": "My org name",
  "ebs_account_id": "2222222",
  "created_at": "2019-02-15T20:26:12.542449Z",
  "updated_at": "2023-01-11T22:16:23.621902Z"
}
========================================
```

#### I have a username. What org do they belong to?

The following command shows all organizations that match the search user string.
```
osdctl org get --user='myuser-like'

sample output
==================================
ID         Name     External ID      EBS ID
111111     ABC      66666            66666
222222     XYZ      77777            77777
333333     PQR      88888            88888
=====================================

osdctl org get --user='myuser-like' -o json

==========================================
{
  "items": [
    {
      "id": "111111",
      "external_id": "66666",
      "name": "ABC",
      "ebs_account_id": "66666",
      "created_at": "2019-10-21T17:22:26.996734Z",
      "updated_at": "2020-11-30T18:00:41.285065Z"
    },
    {
      "id": "111111",
      "external_id": "66666",
      "name": "ABC",
      "ebs_account_id": "66666",
      "created_at": "2019-10-21T17:22:26.996734Z",
      "updated_at": "2020-11-30T18:00:41.285065Z"
    }
  ]
}
==========================================
```
#### I have a partial org name. What org do they belong to?

```
osdctl org get --user='user-like'  --part-match

sample output
==================================
ID         Name     External ID      EBS ID
111111     ABC      66666            66666
222222     XYZ      77777            77777
333333     PQR      88888            88888
=====================================

osdctl org get --user='myuser-like'  --part-match  -o json 

==========================================
{
  "items": [
    {
      "id": "111111",
      "external_id": "66666",
      "name": "ABC",
      "ebs_account_id": "66666",
      "created_at": "2019-10-21T17:22:26.996734Z",
      "updated_at": "2020-11-30T18:00:41.285065Z"
    },
    {
      "id": "111111",
      "external_id": "66666",
      "name": "ABC",
      "ebs_account_id": "66666",
      "created_at": "2019-10-21T17:22:26.996734Z",
      "updated_at": "2020-11-30T18:00:41.285065Z"
    }
  ]
}
==========================================

```

#### What users are in an org?

```
$osdctl org users <orgid>

sample output
==============================================================
USER                  USER ID                       ROLES
abc@example.com   AAABBBCCC   AuthenticatedUser OrganizationMember 
xyz@example.com   XXXYYYZZZ   AuthenticatedUser OrganizationMember 

==============================================================

$osdctl org users <orgid> -o json

==============================================
{
  "items": [
    {
      "user-name": "abc@example.com ",
      "user-id": "AAABBBCCC",
      "roles": [
         "OrganizationMember",
         "RHUser",
         "AuthenticatedUser"
      ]
    },
    {
      "user-name": "xyz@example.com",
      "user-id": "XXXYYYZZZ",
      "roles": [
         "OrganizationMember",
         "OrganizationAdmin",
         "RHUser",
         "AuthenticatedUser"
      ]
    }
  ]
}
==============================================
```

#### What are the capabilities or labels for my cluster?
```
osdctl org labels <orgid>

sample output
==============================================================
ID             KEY                           VALUE
AAACCC        Class                          Customer
BBBDDD        HasTam                         false

==============================================================

osdctl org labels <orgid> -o json

=======================================================
{
  "items": [
    {
      "id": "AAACCC",
      "key": "Class",
      "value": "true"
    },
    {
      "id": "BBBDDD",
      "key": "Customer",
      "value": "false"
    }
  ]
}
=======================================================

```

#### Describe organization
```
osdctl org describe  <orgid>

sample output 
=======================================
ID:              AAAABBBBCCCC
Name:            My org name
External ID:     111111
EBS ID:          2222222
Created:         2019-02-15T20:26:12.542449Z
Updated:         2022-12-21T01:16:25.796805Z
======================================

osdctl org describe  <orgid> -o json

====================================
{
  "id": "AAAABBBBCCCC",
  "external_id": "111111",
  "name": "My org name",
  "ebs_account_id": "2222222",
  "created_at": "2019-02-15T20:26:12.542449Z",
  "updated_at": "2023-01-11T22:16:23.621902Z"
}
====================================

```

#### I have an org ID. What clusters are running in that org?

```
osdctl org clusters  <orgid>
osdctl org clusters  <orgid> -- active

sample output 
==============================================
DISPLAY NAME                   CLUSTER ID                     STATUS
abc-rosa                       AAABBB123456789                Active
xyz-hshift                     XXXYYY123456789                Active
pqr-ccs                        PPPQQQ123456789                Deprovisioned

=============================================

osdctl org clusters  <orgid> -o json
=============================================
{
  "items": [
    {
      "cluster_id": "AAABBB123456789 ",
      "display_name": "abc-rosa",
      "status": "Deprovisioned"
    },
    {
      "cluster_id": " XXXYYY123456789 ",
      "display_name": "pqr-ccs",
      "status": "Active"
    }
  ]
}
=============================================
```

#### What are the paying customer org IDs?

```
osdctl org customers
osdctl org customers --paying=false

sample output
==================================
ID         OrganizationID     SKU 
111111     ABC             MW0055 
222222     XYZ             MW0056  
333333     PQR             MW0057 
=====================================
```

#### I have an OCM org ID. What customer portal ID or access.redhat.com ID is this associated with?
```
osdctl org get --ebs-id=<ebs-account-id>

sample output
==================================
ID         Name     External ID      EBS ID
111111     ABC      66666            66666
222222     XYZ      77777            77777
333333     PQR      88888            88888
=====================================

osdctl org get --ebs-id=<ebs-account-id> -o json
==========================================
{
  "items": [
    {
      "id": "111111",
      "external_id": "66666",
      "name": "ABC",
      "ebs_account_id": "66666",
      "created_at": "2019-10-21T17:22:26.996734Z",
      "updated_at": "2020-11-30T18:00:41.285065Z"
    },
    {
      "id": "111111",
      "external_id": "66666",
      "name": "ABC",
      "ebs_account_id": "66666",
      "created_at": "2019-10-21T17:22:26.996734Z",
      "updated_at": "2020-11-30T18:00:41.285065Z"
    }
  ]
}
==========================================

```
#### List AWS accounts for the given organization unit id
```
$ osdctl org aws-accounts --aws-profile="<aws-profile>" ou-id="<ou-id>"

$ osdctl org aws-accounts --aws-profile="<aws-profile>" ou-id="<ou-id>" -o json
```

### CODE REVIEW

**Comment by supreeth7** [comment]
@samanthajayasinghe can you please add a description?
You can refer https://github.com/openshift/osdctl/pull/296 to get an idea.

File: cmd/org/common.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+		return nil, fmt.Errorf("cannot send request: %q", err)
+	}
+	return response, nil
+}
+
+func checkOrgId(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		err := cmd.Help()
+		if err != nil {
+			return fmt.Errorf("error calling cmd.Help(): %w", err)
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply cmd/org/common.go]: done

File: cmd/org/common.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	}
+	if len(args) != 1 {
+		return fmt.Errorf("Too many arguments. Expected 1 got %d", len(args))
+	}
+
+	return nil
+}
+
+func printOrg(org Organization) (err error) {
+	// Print org details
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply cmd/org/common.go:51]: done

File: cmd/org/common.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		return fmt.Errorf("organization id was not provided. please provide a organization id")
+	}
+	if len(args) != 1 {
+		return fmt.Errorf("Too many arguments. Expected 1 got %d", len(args))
+	}
+
+	return nil
+}
+
+func printOrg(org Organization) (err error) {
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply cmd/org/common.go]: done

File: cmd/org/common.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	// Print org details
+	fmt.Println("ID:  \t \t ", org.ID)
+	fmt.Println("Name: \t \t ", org.Name)
+	fmt.Println("External ID:  \t ", org.ExternalID)
+	fmt.Println("EBS ID:  \t ", org.EBSAccoundID)
+	fmt.Println("Created:  \t ", org.Created)
+	fmt.Println("Updated:  \t ", org.Updated)
+	return nil
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply cmd/org/common.go]: done

File: cmd/org/labels.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+	if err != nil {
+		log.Fatalf("Can't parse API path '%s': %v\n", labelsApiPath, err)
+
+	}
+
+	return request
+}
+
+func printLabels(items []Label) (err error) {
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply cmd/org/labels.go]: done

File: cmd/org/common.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	return response, nil
+}
+
+func checkOrgId(cmd *cobra.Command, args []string) error {
+	if len(args) == 0 {
+		err := cmd.Help()
+		if err != nil {
+			return err
+		}
```
Comment (supreeth7) [inline cmd/org/common.go:40]:
```suggestion
```
We wouldn't want to print the help message for such scenarios.

File: cmd/org/current.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+func run(cmd *cobra.Command) error {
+	response, err := GetCurrentOrg()
+	if err != nil {
+		// If the response has errored, likely the input was bad, so show usage
+		err := cmd.Help()
+		if err != nil {
+			return err
+		}
+		return err
```
Comment (supreeth7) [inline cmd/org/current.go:37]:
```suggestion
		return fmt.Errorf("Invalid input", err)
```

File: cmd/org/current.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	return sendRequest(CreateGetCurrentOrgRequest(ocmClient))
+}
+
+func CreateGetCurrentOrgRequest(ocmClient *sdk.Connection) *sdk.Request {
+	// Create and populate the request:
+	request := ocmClient.Get()
+	err := arguments.ApplyPathArg(request, currentAccountApiPath)
+	if err != nil {
+		log.Fatalf("Can't parse API path '%s': %v\n", currentAccountApiPath, err)
+
```
Comment (supreeth7) [inline cmd/org/current.go:66]:
```suggestion
```

File: cmd/org/describe.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+func DescribeOrg(cmd *cobra.Command, orgID string) error {
+
+	response, err := SendDescribeOrgRequest(orgID)
+	if err != nil {
+		// If the response has errored, likely the input was bad, so show usage
+		err := cmd.Help()
+		if err != nil {
+			return err
+		}
+		return err
```
Comment (supreeth7) [inline cmd/org/describe.go:37]:
```suggestion
		return fmt.Errorf("Invalid input", err)
```

File: cmd/org/describe.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+func CreateDescribeRequest(ocmClient *sdk.Connection, orgID string) *sdk.Request {
+	// Create and populate the request:
+	request := ocmClient.Get()
+	apiPath := organizationsAPIPath + orgID
+
+	err := arguments.ApplyPathArg(request, apiPath)
+
+	if err != nil {
+		log.Fatalf("Can't parse API path '%s': %v\n", apiPath, err)
+
```
Comment (supreeth7) [inline cmd/org/describe.go:70]:
```suggestion
```

File: cmd/org/get.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+)
+
+var (
+	getCmd = &cobra.Command{
+		Use:           "get",
+		Short:         "get oraganization by users",
+		Args:          cobra.ArbitraryArgs,
+		SilenceErrors: true,
+		Run: func(cmd *cobra.Command, args []string) {
+
```
Comment (supreeth7) [inline cmd/org/get.go:23]:
```suggestion
```

File: cmd/org/current.go (hunk 7 above)
Comment (samanthajayasinghe) [reply cmd/org/current.go]: done

File: cmd/org/cmd.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+			err := cmd.Help()
+			if err != nil {
+				fmt.Println("Error calling cmd.Help(): ", err.Error())
+				return
+			}
+		},
+	}
+
+	// Add subcommands
+
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply cmd/org/cmd.go:23]: done

File: cmd/org/common.go [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"fmt"
+
+	sdk "github.com/openshift-online/ocm-sdk-go"
+	"github.com/spf13/cobra"
+)
+
+const (
+	organizationsAPIPath  = "/api/accounts_mgmt/v1/organizations/"
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply cmd/org/common.go]: done

File: cmd/org/common.go (hunk 6 above)
Comment (samanthajayasinghe) [reply cmd/org/common.go]: done

File: cmd/org/current.go (hunk 8 above)
Comment (samanthajayasinghe) [reply cmd/org/current.go]: done

File: cmd/org/describe.go (hunk 9 above)
Comment (samanthajayasinghe) [reply cmd/org/describe.go]: done

File: cmd/org/describe.go (hunk 10 above)
Comment (samanthajayasinghe) [reply cmd/org/describe.go]: done

File: cmd/org/labels.go [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+
+	return nil
+}
+
+func GetLabels(orgID string) (*sdk.Response, error) {
+	// Create OCM client to talk
+	ocmClient := utils.CreateConnection()
+	defer func() {
+		if err := ocmClient.Close(); err != nil {
+			fmt.Printf("Cannot close the ocmClient (possible memory leak): %q", err)
```
[the remark this answers is not in the document]
Comment (samanthajayasinghe) [reply cmd/org/labels.go:60]: the errors are captured from the caller function and the `defer` function may print the errors

File: cmd/org/get.go (hunk 11 above)
Comment (samanthajayasinghe) [reply cmd/org/get.go]: done

**Comment by supreeth7** [comment]
Tested the commands and they all look good.
Please resolve the review comments.

/lgtm

**Comment by iamkirkbater** [comment]
Overall I think this looks great.  Do we think there's any benefit here of also having the ability to print out to JSON via another `-o json` or `--output json` flag?  Though, I don't want to hold this up entirely because of this question.

Hopefully I'll be able to circle back and take a deeper look at the code later this afternoon and give an actual +1 or not later.

**Comment by samanthajayasinghe** [comment]
Thank you very much for checking the PR @iamkirkbater !
It's an excellent point for supporting both `json` and `text` for outputs.
Currently, all org commands output as `text` format. I'll alter all `print***()`  to support json format either in this PR or as a new PR.

**Comment by samanthajayasinghe** [comment]
Hi @iamkirkbater ,
I added `json` output support for all `org` subcommands.  By default, the subcommand lists as text format but `-o json / --output json` will output as JSON format.

**Pull request merged**